File: server/services/footballHints.ts
import { CharacterEntry, HintInfo } from '../types';
import { getThemeConfig } from './themeConfig';

type AllowedClubLeague = { id: string; label: string; name: string };

const footballConfig = getThemeConfig('football');
const allowedClubLeagues =
  (footballConfig.allowedClubLeagues as AllowedClubLeague[] | undefined) ?? [];
const confederations = (footballConfig.confederations as Record<string, string>) ?? {};
const clubLeagues = (footballConfig.clubLeagues as Record<string, string>) ?? {};
const allowedClubLeagueLabels = new Set(allowedClubLeagues.map((l) => l.label));

const HINT_EXCLUDE = new Set([
  'name',
  'id',
  'aliases',
  'displayName',
  'imageUrl',
  'englishName',
  'xhsPlayerId',
  'clubLeague',
]);

export function getFootballHintFields(): string[] {
  return ['club', 'nationalTeam', 'age', 'marketValue', 'height', 'position'];
}

export function getConfederationHint(nationalTeam: unknown): string | null {
  const team = String(nationalTeam || '').trim();
  if (!team) return null;
  return confederations[team] || null;
}

export function resolveClubLeague(entry: CharacterEntry): string | null {
  if (entry.clubLeague && typeof entry.clubLeague === 'string') {
    return entry.clubLeague;
  }
  const club = String(entry.club || '').trim();
  if (!club || club === '无') return null;
  return clubLeagues[club] || null;
}

export function getClubLeagueHint(club: unknown, entry?: CharacterEntry): string | null {
  if (entry) return resolveClubLeague(entry);
  const name = String(club || '').trim();
  if (!name || name === '无') return null;
  return clubLeagues[name] || null;
}

export function isAllowedClubLeague(league: string | null | undefined): boolean {
  if (!league) return false;
  if (allowedClubLeagueLabels.size === 0) return true;
  return allowedClubLeagueLabels.has(league);
}

export function hasValidFootballPrimaryHint(entry: CharacterEntry): boolean {
  return isPlayableFootballAnswer(entry);
}

/**
 * 足球答案池（`filterAtQuestionTime`）：仅 config.allowedClubLeagues 内联赛球员。
 * 与首提示无关——首提示在已选答案上另随机联赛或足联（见 buildFootballPrimaryHint）。
 */
export function isPlayableFootballAnswer(entry: CharacterEntry): boolean {
  return isAllowedClubLeague(resolveClubLeague(entry));
}

function pickRandomIndex(length: number, rng?: { next(): number }): number {
  const r = rng?.next() ?? Math.random();
  return Math.floor(r * length);
}

/** 首提示：在已选答案上，联赛与足联等概率二选一（两者皆有则 50/50）。 */
export function buildFootballPrimaryHint(
  answer: CharacterEntry,
  rng?: { next(): number },
  forcedField?: 'confederation' | 'clubLeague'
): HintInfo {
  const conf = getConfederationHint(answer.nationalTeam);
  const league = getClubLeagueHint(answer.club, answer);

  const options: HintInfo[] = [];
  if (league && isAllowedClubLeague(league)) {
    options.push({ field: 'clubLeague', label: '所属联赛', value: league });
  }
  if (conf) {
    options.push({ field: 'confederation', label: '所属足联', value: conf });
  }

  if (forcedField) {
    const found = options.find((o) => o.field === forcedField);
    if (found) return found;
  }

  if (options.length === 0) {
    return { field: 'confederation', label: '所属足联', value: '未知足联球员' };
  }
  if (options.length === 1) return options[0];
  return options[pickRandomIndex(options.length, rng)];
}

export function buildFootballSyntheticHint(
  answer: CharacterEntry,
  field: 'confederation' | 'clubLeague'
): HintInfo {
  return buildFootballPrimaryHint(answer, undefined, field);
}

export function isFootballHintField(field: string): boolean {
  return field === 'confederation' || field === 'clubLeague';
}

export function isHintFieldExcluded(field: string): boolean {
  return HINT_EXCLUDE.has(field);
}

File: server-go/internal/services/football_hints.go
package services

import (
	"fmt"
	"strings"

	"guess-who/server-go/internal/types"
)

func loadFootballConfig() (confederations, clubLeagues map[string]string, allowedLabels map[string]bool) {
	doc, err := LoadTheme(types.ThemeFootball)
	if err != nil {
		return nil, nil, nil
	}
	confederations = mapFromConfig(doc.Config, "confederations")
	clubLeagues = mapFromConfig(doc.Config, "clubLeagues")
	allowedLabels = make(map[string]bool)
	if raw, ok := doc.Config["allowedClubLeagues"].([]interface{}); ok {
		for _, item := range raw {
			if m, ok := item.(map[string]interface{}); ok {
				if label, ok := m["label"].(string); ok {
					allowedLabels[label] = true
				}
			}
		}
	}
	return confederations, clubLeagues, allowedLabels
}

func mapFromConfig(cfg map[string]interface{}, key string) map[string]string {
	raw, _ := cfg[key].(map[string]interface{})
	out := make(map[string]string, len(raw))
	for k, v := range raw {
		out[k] = fmt.Sprintf("%v", v)
	}
	return out
}

func GetFootballHintFields() []string {
	return []string{"club", "nationalTeam", "age", "marketValue", "height", "position"}
}

func GetConfederationHintForEntry(entry types.CharacterEntry) string {
	_, confederations, _ := loadFootballConfig()
	return GetConfederationHint(fmt.Sprintf("%v", entry["nationalTeam"]), confederations)
}

func ResolveClubLeagueEntry(entry types.CharacterEntry) string {
	_, clubLeagues, _ := loadFootballConfig()
	if v, ok := entry["clubLeague"].(string); ok && strings.TrimSpace(v) != "" {
		return v
	}
	club := strings.TrimSpace(fmt.Sprintf("%v", entry["club"]))
	if club == "" || club == "无" {
		return ""
	}
	return ResolveClubLeague(club, clubLeagues)
}

func GetClubLeagueHintForEntry(entry types.CharacterEntry) string {
	return ResolveClubLeagueEntry(entry)
}

func IsAllowedClubLeagueLabel(league string) bool {
	_, _, allowed := loadFootballConfig()
	if len(allowed) == 0 {
		return league != ""
	}
	return allowed[league]
}

func IsPlayableFootballAnswer(entry types.CharacterEntry) bool {
	league := ResolveClubLeagueEntry(entry)
	return IsAllowedClubLeagueLabel(league)
}

func IsFootballHintField(field string) bool {
	return field == "confederation" || field == "clubLeague"
}

func IsHintFieldExcluded(field string) bool {
	excluded := map[string]bool{
		"name": true, "id": true, "aliases": true, "displayName": true,
		"imageUrl": true, "englishName": true, "xhsPlayerId": true, "clubLeague": true,
	}
	return excluded[field]
}

func BuildFootballPrimaryHint(answer types.CharacterEntry, rng *SeededRng, forcedField string) types.HintInfo {
	confederations, _, _ := loadFootballConfig()
	conf := GetConfederationHint(fmt.Sprintf("%v", answer["nationalTeam"]), confederations)
	league := GetClubLeagueHintForEntry(answer)

	var options []types.HintInfo
	if league != "" && IsAllowedClubLeagueLabel(league) {
		options = append(options, types.HintInfo{Field: "clubLeague", Label: "所属联赛", Value: league})
	}
	if conf != "" {
		options = append(options, types.HintInfo{Field: "confederation", Label: "所属足联", Value: conf})
	}

	if forcedField != "" {
		for _, o := range options {
			if o.Field == forcedField {
				return o
			}
		}
	}

	if len(options) == 0 {
		return types.HintInfo{Field: "confederation", Label: "所属足联", Value: "未知足联球员"}
	}
	if len(options) == 1 {
		return options[0]
	}
	idx := 0
	if rng != nil {
		idx = rng.NextInt(len(options))
	}
	return options[idx]
}

func BuildFootballSyntheticHint(answer types.CharacterEntry, field string) types.HintInfo {
	return BuildFootballPrimaryHint(answer, nil, field)
}

// GetConfederationHint 返回国家队对应的洲际赛区。
func GetConfederationHint(nationalTeam string, confederations map[string]string) string {
	team := strings.TrimSpace(nationalTeam)
	if team == "" || confederations == nil {
		return ""
	}
	if v, ok := confederations[team]; ok {
		return v
	}
	return ""
}

// ResolveClubLeague 从配置查找俱乐部对应的联赛。
func ResolveClubLeague(club string, clubLeagues map[string]string) string {
	c := strings.TrimSpace(club)
	if c == "" || clubLeagues == nil {
		return ""
	}
	if league, ok := clubLeagues[c]; ok {
		return league
	}
	return ""
}

func IsAllowedClubLeague(league string, allowed map[string]bool) bool {
	if allowed == nil || len(allowed) == 0 {
		return true
	}
	return allowed[league]
}
